geoh5py/shared/merging/drape_model.py

```diff
@@ -0,0 +1,180 @@
+                layers.append(ghost_layer)
+
+            # add the entity prisms and layers to the list
+            temp_prisms[:, -2] += previous_prism
+            temp_layers[:, 0] += previous_layer
+
+            previous_prism = temp_prisms[-1, -1] + temp_prisms[-1, -2] + 2
+            previous_layer = temp_layers[-1, 0] + 3
+
+            prisms.append(temp_prisms)
+            layers.append(temp_layers)
+
+            # create a last ghost prism and layer
+            ghost_prism, ghost_layer = cls._ghost_point(
+                temp_prisms[-1], temp_prisms[-2], previous_prism - 2, previous_layer - 2
+            )
+
+        # create an object
+        output = cls._type.create(  # type: ignore
+            workspace, prisms=np.vstack(prisms), layers=np.vstack(layers), **kwargs
+        )
+
+        return output
+
+    @staticmethod
+    def _ghost_point(
+        point: np.ndarray,
+        mirror: np.ndarray,
+        previous_prism: int,
+        previous_layer: int,
+    ) -> np.ndarray | np.ndarray:
+        """
+        Create a ghost point prism and layer based on two points.
+        :param point: The point to create the ghost from.
+        :param mirror: The mirror point to create the ghost from.
+        :param previous_prism: The ID of the previous prism.
+        :param previous_layer: The ID of the previous layer.
+        :return: A prism and a layer for the ghost point.
+        """
+
+        # create a mirrored point for last layer
+        ghost_prism: np.ndarray = point.copy()
+        ghost_layer: np.ndarray = np.empty(3)
+        ghost_prism[:3] = 2 * point[:3] - mirror[:3]
+
+        ghost_prism[3] = previous_prism
+        ghost_prism[4] = 1
+
+        ghost_layer[0] = previous_layer
+        ghost_layer[1] = 0
+        ghost_layer[2] = ghost_prism[2]
+
+        return np.expand_dims(ghost_prism, 0), np.expand_dims(ghost_layer, 0)
+
+    @classmethod
+    def merge_data(
+        cls,
+        out_entity,
+        input_entities: list[ObjectBase],
+    ):
+        super().merge_data(out_entity, input_entities)
+
+        # prepare an empty nan array
+        values: np.ndarray = np.empty(out_entity.n_cells)
+        values[:] = np.nan
+
+        # get all the values in the output entity
+        for data in out_entity.children:
+            if (
+                not isinstance(data, NumericData)
+                or data.association is None
+                or data.n_values is None
+                or data.values is None
+            ):
+                continue
+            previous: int = 0
+            twos: int = 0
+            for input_entity in input_entities:
+                n_cells = cast(int, input_entity.n_cells)
+                values[previous + twos : previous + twos + n_cells] = data.values[
+                    previous : previous + n_cells
+                ]
+
+                previous += n_cells
+                twos += 2
+
+            # update the values
+            data.values = values
```

Comment:
You could just create an index map once, and apply on all data. Will be faster than doing to nested loops.